phoenix-core/src/it/java/org/apache/phoenix/end2end/DeleteIT.java

```diff
@@ -943,6 +960,164 @@ public void testDeleteFilterWithMultipleIndexes() throws Exception {
+            assertTrue(plan.getClass().getName().contains(planName));
+        }
+    }
+    private void createAndUpsertTable(String tableName, String indexName, Properties props)
+            throws SQLException {
+        String ddl = "CREATE TABLE " + tableName +
+                " (k INTEGER NOT NULL PRIMARY KEY, v1 VARCHAR, v2 VARCHAR)";
+        try(Connection conn = DriverManager.getConnection(getUrl(), props)) {
+            conn.setAutoCommit(true);
+            try (Statement statement = conn.createStatement()) {
+                statement.execute(ddl);
+                if (indexName != null) {
+                    String indexDdl1 = "CREATE INDEX " + indexName + " ON " + tableName + "(v1,v2)";
+                    statement.execute(indexDdl1);
+                }
+            }
+            conn.createStatement().execute(
+                    "upsert into " + tableName + " values (1, 'foo', 'foo1')");
+            conn.createStatement().execute(
+                    "upsert into " + tableName + " values (2, 'bar', 'bar1')");
+            conn.commit();
+        }
+    }
+
+    private void executeDelete(String delete, Properties props, int deleteRowCount)
+            throws SQLException {
+        try(Connection conn = DriverManager.getConnection(getUrl(), props)) {
+            conn.setAutoCommit(true);
+            try (Statement statement = conn.createStatement()) {
+                int rs = statement.executeUpdate(delete);
+                assertEquals( deleteRowCount, rs);
+            }
+        }
+    }
 
+    /*
+        Verify whether we have tags present for base table and not present for
+        index tables.
+     */
+    private void checkTagPresentInDeleteMarker(String tableName, String startRowKey,
+            boolean tagPresent, String tagValue) throws IOException {
+        List<Cell> values = new ArrayList<>();
+        TableName table = TableName.valueOf(tableName);
+        // Scan table with specified startRowKey
+        for (HRegion region : getUtility().getHBaseCluster().getRegions(table)) {
+            values.clear();
+            Scan scan = new Scan();
+            // Make sure to set rawScan to true so that we will get Delete Markers.
+            scan.setRaw(true);
+            scan.withStartRow(Bytes.toBytes(startRowKey));
+            RegionScanner scanner = region.getScanner(scan);
+            scanner.next(values);
+            if (!values.isEmpty()) {
+                break;
+            }
+        }
+        assertTrue("Values shouldn't be empty", !values.isEmpty());
+        Cell first = values.get(0);
+        assertTrue("First cell should be delete marker ",
+                PrivateCellUtil.isDelete(first.getType().getCode()));
+        List<Tag> tags = PrivateCellUtil.getTags(first);
+        if (tagPresent) {
+            assertEquals(1, tags.size());
+            Optional<Tag> optional =
+                    PrivateCellUtil.getTag(first, PhoenixTagType.SOURCE_OPERATION_TAG_TYPE);
```

Comment:
PrivateCellUtil is IA.Private -- need to use one of the IA.LimitedPrivate or better classes such as RawCell